x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -173,7 +173,7 @@ public List<Gap> waitForRange(final ByteRange range, final ByteRange subRange, f
             );
         }
 
-        if (complete >= range.end()) {
+        if (subRange.end() <= complete) {
```

Comment:
Can we make this clear in the javadoc, i.e., if the subrange is already present we may not fill the range.

I think this may not mean a lot in practice in that we already do the `tryRead` which should catch this except for race conditions (and there is not that far between the two checks). I am good with it anyway, seems right to me.

Reply:
I pushed 583ff31a2417c13771aa2c399a56de20170a8cf3 for the doc.

> I think this may not mean a lot in practice in that we already do the tryRead which should catch this except for race conditions (and there is not that far between the two checks). I am good with it anyway, seems right to me.

Yes, it seemed more correct to me too.